File: src/main/java/io/github/seggan/jyxal/runtime/ProgramStack.java
package io.github.seggan.jyxal.runtime;

import io.github.seggan.jyxal.runtime.math.BigComplex;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayDeque;
import java.util.Collection;
import java.util.Deque;
import java.util.List;

public class ProgramStack extends ArrayDeque<Object> implements Deque<Object> {

    private Object[] input;
    private String flags;

    private int index;

    public ProgramStack() {
        super();
        this.input = null;
        this.flags = null;
    }

    // the String[] is the program args
    public ProgramStack(String[] strings) throws IOException {
        super();
        if (strings.length > 0) {
            this.flags = strings[0];
            if (strings.length > 1) {
                if (this.flags.indexOf('f') != -1) {
                    this.input = new Object[1];
                    this.input[0] = Files.readString(Path.of(strings[1]));
                } else {
                    this.input = new Object[strings.length - 1];
                    for (int i = 1; i < strings.length; i++) {
                        this.input[i - 1] = RuntimeHelpers.eval(strings[i]);
                    }
                }
            }
        }
    }

    public ProgramStack(Object... objects) {
        super(List.of(objects));
        this.input = objects;
        this.flags = null;
    }

    public ProgramStack(Collection<?> c) {
        super(c);
        this.input = c.toArray();
        this.flags = null;
    }

    @Override
    public Object pop() {
        if (this.isEmpty()) {
            return getInput();
        } else {
            return super.pop();
        }
    }

    public void swap() {
        Object a = this.pop();
        Object b = this.pop();
        this.push(a);
        this.push(b);
    }

    /**
     * This modifies this stack
     */
    public ProgramStack reverse() {
        ProgramStack reversed = new ProgramStack();
        while (!this.isEmpty()) {
            reversed.push(this.pop());
        }
        reversed.input = this.input;
        reversed.flags = this.flags;
        return reversed;
    }

    public void push(boolean b) {
        push(BigComplex.valueOf(b));
    }

    public void push(long i) {
        push(BigComplex.valueOf(i));
    }

    public Object getInput() {
        if (input == null || input.length == 0) {
            return BigComplex.ZERO;
        } else {
            index %= input.length;
            return input[index++];
        }
    }
}

File: src/main/kotlin/io/github/seggan/jyxal/runtime/ProgramStack.kt
package io.github.seggan.jyxal.runtime

import io.github.seggan.jyxal.runtime.math.BigComplex
import java.nio.file.Files
import java.nio.file.Path
import java.util.ArrayDeque
import java.util.Deque

class ProgramStack : ArrayDeque<Any?>, Deque<Any?> {

    private var input: Array<Any>?
    private var flags: String? = null
    private var index = 0

    constructor() : super() {
        input = null
        flags = null
    }

    // the String[] is the program args
    constructor(strings: Array<String>) : super() {
        if (strings.isNotEmpty()) {
            flags = strings[0]
            if (strings.size > 1) {
                if (flags!!.indexOf('f') != -1) {
                    input = arrayOf(1)
                    input!![0] = Files.readString(Path.of(strings[1]))
                } else {
                    input = arrayOf(strings.size - 1)
                    for (i in 1 until strings.size) {
                        input!![i - 1] = eval(strings[i])
                    }
                }
            }
        }
        input = null
    }

    constructor(vararg objects: Any?) : super(listOf(*objects)) {
        input = arrayOf(objects)
        flags = null
    }

    constructor(c: Collection<Any>) : super(c) {
        input = c.toTypedArray()
        flags = null
    }

    override fun pop(): Any {
        @Suppress("UNNECESSARY_NOT_NULL_ASSERTION")
        return if (this.isEmpty()) {
            getInput()
        } else {
            super.pop()!!
        }
    }

    fun swap() {
        val a = pop()
        val b = pop()
        this.push(a)
        this.push(b)
    }

    /**
     * This modifies this stack
     */
    fun reverse(): ProgramStack {
        val reversed = ProgramStack()
        while (!this.isEmpty()) {
            reversed.push(pop())
        }
        reversed.input = input
        reversed.flags = flags
        return reversed
    }

    fun push(b: Boolean) {
        push(BigComplex.valueOf(b))
    }

    fun push(i: Long) {
        push(BigComplex.valueOf(i))
    }

    fun getInput(): Any {
        return if (input == null || input!!.isEmpty()) {
            BigComplex.ZERO
        } else {
            index %= input!!.size
            input!![index++]
        }
    }
}